logtail/logtail_test.go

```diff
@@ -23,29 +28,245 @@ func TestFastShutdown(t *testing.T) {
 	l := NewLogger(Config{
 		BaseURL: testServ.URL,
 	}, t.Logf)
-	l.Shutdown(ctx)
+	err := l.Shutdown(ctx)
+	if err != nil {
+		t.Error(err)
+	}
 }
 
-func TestUploadMessages(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
-	uploads := 0
-	testServ := httptest.NewServer(http.HandlerFunc(
+type LogtailTestServer struct {
+	ctx      context.Context
+	cancel   context.CancelFunc
+	uploaded chan int
+	logLines int // number of messages to be received
+	body     []byte
+	bodytext string
+	mu       sync.Mutex
+	srv      *httptest.Server
+}
+
+func NewLogtailTestServer(logLines int) *LogtailTestServer {
+	ts := LogtailTestServer{logLines: logLines}
+
+	// max channel backlog = 1 "started" + logLines x "log line" + 1 "closed"
+	ts.uploaded = make(chan int, 2+logLines)
+
+	ts.ctx, ts.cancel = context.WithCancel(context.Background())
+
+	ts.srv = httptest.NewServer(http.HandlerFunc(
 		func(w http.ResponseWriter, r *http.Request) {
-			w.Header().Set("Content-Type", "application/json; charset=utf-8")
-			uploads += 1
+			body, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				panic("failed to read HTTP request")
+			}
+			ts.mu.Lock()
+			defer ts.mu.Unlock()
+			ts.body = body
+			ts.bodytext += string(body) + "\n"
+			ts.uploaded <- 0
 		}))
-	defer testServ.Close()
 
-	l := NewLogger(Config{BaseURL: testServ.URL}, t.Logf)
-	for i := 1; i < 10; i++ {
+	return &ts
+}
+
+func (ts *LogtailTestServer) HandleStartLoggingMessage() {
+	// There is always an initial "logtail started" message
+	<-ts.uploaded
+	ts.mu.Lock()
+	if !strings.Contains(ts.bodytext, "started") {
+		panic("unknown start logging statement:" + ts.bodytext)
+	}
+	ts.mu.Unlock()
+}
+
+func TestDrainPendingMessages(t *testing.T) {
+	logLines := 3
+	ts := NewLogtailTestServer(logLines)
+	l := NewLogger(Config{BaseURL: ts.srv.URL}, t.Logf)
+	ts.HandleStartLoggingMessage()
+
+	for i := 0; i < logLines; i++ {
 		l.Write([]byte("log line"))
 	}
 
-	l.Shutdown(ctx)
-	cancel()
-	if uploads == 0 {
-		t.Error("no log uploads")
+	// all of the "log line" messages usually arrive at once, but poll if needed.
+	for i := 0; i <= logLines; i++ {
+		<-ts.uploaded
+		ts.mu.Lock()
+		cnt := strings.Count(ts.bodytext, "log line")
+		ts.mu.Unlock()
+		if cnt == logLines {
+			break
+		}
 	}
+
+	err := l.Shutdown(ts.ctx)
```

Comment:
Can you use `context.Background()` here? If so, then I think you can get rid of LogtailTestServer.{ctx,cancel}.

Reply:
I recall that being able to call cancel() sped up running the test in a loop, but it no longer does. Removed {ctx, cancel} from struct and used context.Background() for shutdown.